packages/video_player/video_player_web/lib/video_player_web.dart

```diff
@@ -144,10 +157,453 @@ class VideoPlayerPlugin extends VideoPlayerPlatform {
+  /// while optimizing performance. If using HTML renderer, will use a platform
+  /// view.
+  auto
+}
+
+abstract class _VideoPlayerRenderer extends StatefulWidget {
+  const _VideoPlayerRenderer({required this.element, this.paused = true});
+
+  final web.HTMLVideoElement element;
+  final bool paused;
+}
+
+abstract class _VideoPlayerRendererState<T extends _VideoPlayerRenderer>
+    extends State<T> {
+  _VideoPlayerRendererState() {
+    onSeekRef = onSeek.toJS;
+  }
+
+  @override
+  void initState() {
+    super.initState();
+
+    frameCallback(0.toJS, 0.toJS);
+    widget.element.addEventListener('seeked', onSeekRef);
+  }
+
+  late JSFunction onSeekRef;
+
+  int? callbackID;
+
+  void getFrame(web.HTMLVideoElement element) {
+    callbackID =
+        element.requestVideoFrameCallbackWithFallback(frameCallback.toJS);
+  }
+
+  void cancelFrame(web.HTMLVideoElement element) {
+    if (callbackID != null) {
+      element.cancelVideoFrameCallbackWithFallback(callbackID!);
+    }
+  }
+
+  void onSeek(web.Event event) {
+    capture();
+  }
+
+  void frameCallback(JSAny now, JSAny metadata) {
+    final web.HTMLVideoElement element = widget.element;
+    final bool isPlaying = !!(element.currentTime > 0 &&
+        !element.paused &&
+        !element.ended &&
+        element.readyState > 2);
+
+    // only capture frames if video is playing (optimization for RAF)
+    if (isPlaying || element.readyState > 2 && lastFrameTime == null) {
+      capture().then((_) async {
+        getFrame(widget.element);
+      });
+    } else {
+      getFrame(widget.element);
+    }
+  }
+
+  web.ImageBitmap? source;
+  num? lastFrameTime;
+
+  Future<void> capture();
+
+  @override
+  void dispose() {
+    cancelFrame(widget.element);
+    super.dispose();
+    source?.close();
+    widget.element.removeEventListener('seeked', onSeekRef);
+  }
+
+  @override
+  void didUpdateWidget(T oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.element != widget.element) {
+      oldWidget.element.addEventListener('seeked', onSeekRef);
+      widget.element.addEventListener('seeked', onSeekRef);
+      cancelFrame(oldWidget.element);
+      getFrame(widget.element);
+    }
+  }
+}
+
+class _TextureVideoPlayerRenderer extends _VideoPlayerRenderer {
+  const _TextureVideoPlayerRenderer({required super.element, super.paused});
+
+  @override
+  State createState() => _TextureVideoPlayerRendererState();
+}
+
+class _TextureVideoPlayerRendererState
+    extends _VideoPlayerRendererState<_TextureVideoPlayerRenderer> {
+  @override
+  Future<void> capture() async {
+    if (!widget.paused || lastFrameTime != widget.element.currentTime) {
+      lastFrameTime = widget.element.currentTime;
+      try {
+        final web.ImageBitmap newSource =
+            await promiseToFuture<web.ImageBitmap>(
+                web.window.createImageBitmap(widget.element));
+        final ui.Image img = await ui_web.createImageFromImageBitmap(newSource);
```

Comment:
This

Reply:
@ditman It's definitely not the fastest way the browser supports, but it is the only way that's currently exposed to get a frame into the renderer. I have some suggestions here in this thread:

https://github.com/flutter/flutter/issues/144815#issuecomment-1987309476

Ideally, in browsers where a VideoFrame is available, it could be transferred to the worker for rendering. There are some constraints with VideoFrames though that you can't hold onto them without blocking the decoding, so might require some additional plumbing to make work properly. Maybe WebCodecs would be the best option at some point: https://github.com/w3c/webcodecs/issues/43. But also not doable without some work in the flutter engine.